## Opening Upscale Italian Restaurants in Los Angeles


Company A is planning to open several upscale Italian restaurants in Los Angeles.

They are deciding on the areas, the criteria are:
the areas should have a high median income
the areas should not have many Italian restaurants around it
the areas should have a fairly large population size

The solution is provided using the KMeans clustering method


### Importing dataset

The dataset is retrieved from Los Angeles Almanac (http://www.laalmanac.com/employment/em12c.php) and ZIPatlas.com (http://zipatlas.com/us/ca/los-angeles/zip-code-comparison/population-density.htm)



In [154]:
import pandas as pd

la_data = pd.read_csv('la_data.csv')

la_data['Zip - Neighborhood'] = la_data['Zip'].map(str)+' - '+la_data['Neighborhood']

arrange_columns = [la_data.columns[-1]] + list(la_data.columns[:-1])
la_data = la_data[arrange_columns]
la_data.columns = ['Neighborhood','Zip','Name','Latitude','Longitude','Population','Median Income']
la_data

Neighborhood    Zip  \
0                       90057 - Los Angeles(Westlake)  90057   
1   90020 - Los Angeles(Hancock Park, Western Wilt...  90020   
2   90005 - Los Angeles(Hancock Park, Koreatown, W...  90005   
3   90006 - Los Angeles(Byzantine-Latino Quarter, ...  90006   
4                 90029 - Los Angeles(East Hollywood)  90029   
5   90017 - Los Angeles(Downtown Bunker Hill, City...  90017   
6          90011 - Los Angeles(Southeast Los Angeles)  90011   
7   90004 - Los Angeles(Hancock Park, Rampart Vill...  90004   
8                      90038 - Los Angeles(Hollywood)  90038   
9                      90028 - Los Angeles(Hollywood)  90028   
10             90037 - Los Angeles(South Los Angeles)  90037   
11                         90034 - Los Angeles(Palms)  90034   
12        90026 - Los Angeles(Echo Park, Silver Lake)  90026   
13  90019 - Los Angeles(Arlington Heights, Country...  90019   
14      90044 - Athens,Los Angeles(South Los Angeles)  90044   
15  90003 - Los Angeles(South Los Angeles, Southea...  90003   
16  90007 - Los Angeles(Southeast Los Angeles, Uni...  90007   
17  90018 - Los Angeles(Jefferson Park, Leimert Park)  90018   
18                 90033 - Los Angeles(Boyle Heights)  90033   
19    90063 - City Terrace,Los Angeles(Boyle Heights)  90063   
20  90001 - Los Angeles(South Los Angeles), Floren...  90001   
21                           90022 - East Los Angeles  90022   
22             90062 - Los Angeles(South Los Angeles)  90062   
23    90025 - Los Angeles(Sawtelle, West Los Angeles)  90025   
24  90002 - Los Angeles(Southeast Los Angeles, Watts)  90002   
25                 90042 - Los Angeles(Highland Park)  90042   
26                  90035 - Los Angeles(West Fairfax)  90035   
27  90036 - Los Angeles(Fairfax, Melrose, Miracle ...  90036   
28                      90024 - Los Angeles(Westwood)  90024   
29                    90016 - Los Angeles(West Adams)  90016   
30  90013 - Los Angeles(Downtown Central, Downtown...  90013   
31  90014 - Los Angeles(Downtown Historic Core, Ar...  90014   
32  90059 - Los Angeles(Southeast Los Angeles, Wat...  90059   
33  90048 - Los Angeles(Mid-City West), West Holly...  90048   
34         90066 - Culver City,Los Angeles(Mar Vista)  90066   
35  90043 - Los Angeles(Hyde Park, View Park, Wind...  90043   
36             90047 - Los Angeles(South Los Angeles)  90047   
37     90032 - Los Angeles(El Sereno, Monterey Hills)  90032   
38  90031 - Los Angeles(Lincoln Heights, Montecito...  90031   
39  90065 - Los Angeles(Cypress Park, Glassell Par...  90065   
40             90061 - Los Angeles(South Los Angeles)  90061   
41  90015 - Los Angeles(Dowtown Fashion District, ...  90015   
42     90046 - Los Angeles(Hollywood), West Hollywood  90046   
43                  90023 - Commerce,East Los Angeles  90023   
44  90012 - Los Angeles(Downtown Civic Center, Chi...  90012   
45  90008 - Los Angeles(Baldwin Hills, Crenshaw, L...  90008   
46                  90067 - Los Angeles(Century City)  90067   
47                    90041 - Los Angeles(Eagle Rock)  90041   
48  90039 - Los Angeles(Atwater Village, Elysian V...  90039   
49    90064 - Los Angeles(Cheviot Hills, Rancho Park)  90064   
50  90027 - Los Angeles(Griffith Park, Hollywood, ...  90027   
51                             90056 - Ladera Heights  90056   
52  90010 - Los Angeles(Hancock Park, Wilshire Cen...  90010   
53  90045 - Los Angeles(Los Angeles International ...  90045   
54                     90068 - Los Angeles(Hollywood)  90068   
55    90049 - Los Angeles(Bel Air Estates, Brentwood)  90049   
56                          90040 - Commerce, City of  90040   
57  90021 - Los Angeles(Downtown Fashion District,...  90021   
58  90077 - Los Angeles(Bel Air Estates, Beverly G...  90077   
59  90058 - Los Angeles(Southeast Los Angeles), Ve...  90058   

                                                 Name   Latitude   Longitude  \
0                               Los Angeles(Westlake)  3

### Importing libraries

In [14]:
import json
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium # map rendering library


Solving environment: done

# All requested packages already installed.



### Creating a map of Los Angeles

In [15]:
latitude = 34.0522
longitude = -118.2437
print('The geograpical coordinate of Los Angeles are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Los Angeles are 34.0522, -118.2437.


In [155]:
map_la = folium.Map(location=[latitude, longitude], zoom_start = 10)
map_la

for lat, lng, neighborhood in zip(la_data['Latitude'], la_data['Longitude'], la_data['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat,lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_la)
    
map_la

### Getting the venues from Foursquare

In [17]:
#Define Foursquare credentials and version

CLIENT_ID = 'F44RE35X5VD0OR4QZ5QFM2VVP5PRMMMB1JXKTFKKAVU5HBN3' 
CLIENT_SECRET = 'FOVYKDZGWC0GURZJTGA3CXCXZLMXZ3IXYWXYU2GD0Z3TMQS1' 
VERSION = '20180605' 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: F44RE35X5VD0OR4QZ5QFM2VVP5PRMMMB1JXKTFKKAVU5HBN3
CLIENT_SECRET:FOVYKDZGWC0GURZJTGA3CXCXZLMXZ3IXYWXYU2GD0Z3TMQS1


In [156]:
#Getting 100 venues within a radius of 1000 meters (1 kilometer)

LIMIT = 100
radius = 1000

def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names,latitudes,longitudes):
        print(name)
        
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        results = requests.get(url).json()["response"]["groups"][0]['items']
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


In [157]:
la_venues = getNearbyVenues(names=la_data['Neighborhood'],
                                   latitudes=la_data['Latitude'],
                                   longitudes=la_data['Longitude']
                                  )

90057 - Los Angeles(Westlake)
90020 - Los Angeles(Hancock Park, Western Wilton, Wilshire Center, Windsor Square)
90005 - Los Angeles(Hancock Park, Koreatown, Wilshire Center, Wilshire Park, Windsor Square)
90006 - Los Angeles(Byzantine-Latino Quarter, Harvard Heights, Koreatown, Pico Heights)
90029 - Los Angeles(East Hollywood)
90017 - Los Angeles(Downtown Bunker Hill, City West, South Park-North)
90011 - Los Angeles(Southeast Los Angeles)
90004 - Los Angeles(Hancock Park, Rampart Village, Virgil Village, Wilshire Center, Windsor Square)
90038 - Los Angeles(Hollywood)
90028 - Los Angeles(Hollywood)
90037 - Los Angeles(South Los Angeles)
90034 - Los Angeles(Palms)
90026 - Los Angeles(Echo Park, Silver Lake)
90019 - Los Angeles(Arlington Heights, Country Club Park, Mid-City)
90044 - Athens,Los Angeles(South Los Angeles)
90003 - Los Angeles(South Los Angeles, Southeast Los Angeles)
90007 - Los Angeles(Southeast Los Angeles, Univerity Park)
90018 - Los Angeles(Jefferson Park, Leimert Park)

In [158]:
la_onehot = pd.get_dummies(la_venues[['Venue Category']], prefix="", prefix_sep="")
la_onehot['_Neighborhood'] = la_venues['Neighborhood']
la_onehot.drop('Neighborhood',axis=1, inplace = True)
fixed_columns = [la_onehot.columns[-1]] + list(la_onehot.columns[:-1])
la_onehot = la_onehot[fixed_columns]
la_onehot = la_onehot.rename(columns={'_Neighborhood':'Neighborhood'})

la_onehot.head()

Neighborhood  ATM  Accessories Store  Airport Lounge  \
0  90057 - Los Angeles(Westlake)    0                  0               0   
1  90057 - Los Angeles(Westlake)    0                  0               0   
2  90057 - Los Angeles(Westlake)    0                  0               0   
3  90057 - Los Angeles(Westlake)    0                  0               0   
4  90057 - Los Angeles(Westlake)    0                  0               0   

   Airport Service  Airport Terminal  American Restaurant  Amphitheater  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Antique Shop  Arcade  ...  Vietnamese Restaurant  Warehouse  \
0             0       0  ...                      0          0   
1             0       0  ...                      0          0   
2             0       0  ...                      0          0   
3             0       0  ...                      0          0   
4             0       0  ...                      0          0   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                0           0         0          0            0   
1                0           0         0          0            0   
2                0           0         0          0            0   
3                0           0         0          0            0   
4                0           0         0          0            0   

   Women's Store  Yoga Studio  Yoshoku Restaurant  
0              0            0                   0  
1              0            0                   0  
2              0            0                   0  
3              0            0                   0  
4              0            0                   0  

[5 rows x 329 columns]

In [159]:
la_types = list(la_onehot.columns)
la_types

['Neighborhood',
 'ATM',
 'Accessories Store',
 'Airport Lounge',
 'Airport Service',
 'Airport Terminal',
 'American Restaurant',
 'Amphitheater',
 'Antique Shop',
 'Arcade',
 'Art Gallery',
 'Art Museum',
 'Arts & Crafts Store',
 'Asian Restaurant',
 'Astrologer',
 'Athletics & Sports',
 'Auto Garage',
 'Automotive Shop',
 'BBQ Joint',
 'Bagel Shop',
 'Bakery',
 'Bank',
 'Bar',
 'Baseball Field',
 'Baseball Stadium',
 'Basketball Court',
 'Basketball Stadium',
 'Bed & Breakfast',
 'Beer Bar',
 'Beer Garden',
 'Beer Store',
 'Big Box Store',
 'Bike Shop',
 'Bookstore',
 'Boutique',
 'Bowling Alley',
 'Boxing Gym',
 'Brazilian Restaurant',
 'Breakfast Spot',
 'Brewery',
 'Bridal Shop',
 'Bubble Tea Shop',
 'Buffet',
 'Burger Joint',
 'Burmese Restaurant',
 'Burrito Place',
 'Bus Line',
 'Bus Station',
 'Bus Stop',
 'Business Service',
 'Cafeteria',
 'Café',
 'Cajun / Creole Restaurant',
 'Campground',
 'Candy Store',
 'Cantonese Restaurant',
 'Caribbean Restaurant',
 'Carpet Store',
 '

In [160]:
#Grouping the venues by Neighborhood

la_grouped = la_onehot.groupby("Neighborhood").sum().reset_index()

la_grouped

Neighborhood  ATM  Accessories Store  \
0   90001 - Los Angeles(South Los Angeles), Floren...    0                  0   
1   90002 - Los Angeles(Southeast Los Angeles, Watts)    0                  0   
2   90003 - Los Angeles(South Los Angeles, Southea...    1                  0   
3   90004 - Los Angeles(Hancock Park, Rampart Vill...    0                  0   
4   90005 - Los Angeles(Hancock Park, Koreatown, W...    0                  0   
5   90006 - Los Angeles(Byzantine-Latino Quarter, ...    0                  0   
6   90007 - Los Angeles(Southeast Los Angeles, Uni...    1                  0   
7   90008 - Los Angeles(Baldwin Hills, Crenshaw, L...    0                  1   
8   90010 - Los Angeles(Hancock Park, Wilshire Cen...    0                  0   
9          90011 - Los Angeles(Southeast Los Angeles)    0                  0   
10  90012 - Los Angeles(Downtown Civic Center, Chi...    0                  0   
11  90013 - Los Angeles(Downtown Central, Downtown...    0                  0   
12  90014 - Los Angeles(Downtown Historic Core, Ar...    0                  0   
13  90015 - Los Angeles(Dowtown Fashion District, ...    0                  0   
14                    90016 - Los Angeles(West Adams)    0                  0   
15  90017 - Los Angeles(Downtown Bunker Hill, City...    0                  0   
16  90018 - Los Angeles(Jefferson Park, Leimert Park)    0                  0   
17  90019 - Los Angeles(Arlington Heights, Country...    0                  0   
18  90020 - Los Angeles(Hancock Park, Western Wilt...    0                  0   
19  90021 - Los Angeles(Downtown Fashion District,...    0                  0   
20                           90022 - East Los Angeles    0                  0   
21                  90023 - Commerce,East Los Angeles    0                  0   
22                      90024 - Los Angeles(Westwood)    0                  0   
23    90025 - Los Angeles(Sawtelle, West Los Angeles)    0                  0   
24        90026 - Los Angeles(Echo Park, Silver Lake)    0                  0   
25  90027 - Los Angeles(Griffith Park, Hollywood, ...    0                  0   
26                     90028 - Los Angeles(Hollywood)    0                  0   
27                90029 - Los Angeles(East Hollywood)    0                  0   
28  90031 - Los Angeles(Lincoln Heights, Montecito...    0                  0   
29     90032 - Los Angeles(El Sereno, Monterey Hills)    1                  0   
30                 90033 - Los Angeles(Boyle Heights)    0                  0   
31                         90034 - Los Angeles(Palms)    0                  0   
32                  90035 - Los Angeles(West Fairfax)    0                  0   
33  90036 - Los Angeles(Fairfax, Melrose, Miracle ...    0                  1   
34             90037 - Los Angeles(South Los Angeles)    0                  0   
35                     90038 - Los Angeles(Hollywood)    0                  0   
36  90039 - Los Angeles(Atwater Village, Elysian V...    0                  0   
37                          90040 - Commerce, City of    0                  0   
38                    90041 - Los Angeles(Eagle Rock)    0                  0   
39                 90042 - Los Angeles(Highland Park)    0                  0   
40  90043 - Los Angeles(Hyde Park, View Park, Wind...    1                  0   
41      90044 - Athens,Los Angeles(South Los Angeles)    0                  0   
42  90045 - Los Angeles(Los Angeles International ...    0                  0   
43     90046 - Los Angeles(Hollywood), West Hollywood    0                  0   
44             90047 - Los Angeles(South Los Angeles)    0                  0   
45  90048 - Los Angeles(Mid-City West), West Holly...    0                  0   
46    90049 - Los Angeles(Bel Air Estates, Brentwood)    0                  0   
47                             90056 - Ladera Heights    0                  0   
48                      90057 - Los Angeles(Westlake)    0              

In [228]:
#Keep only number of Italian Restaurant

la_italian = la_grouped[['Neighborhood','Italian Restaurant']]
la_italian.head()

Neighborhood  Italian Restaurant
0  90001 - Los Angeles(South Los Angeles), Floren...                   0
1  90002 - Los Angeles(Southeast Los Angeles, Watts)                   0
2  90003 - Los Angeles(South Los Angeles, Southea...                   0
3  90004 - Los Angeles(Hancock Park, Rampart Vill...                   1
4  90005 - Los Angeles(Hancock Park, Koreatown, W...                   1

In [184]:
#Merge Italian restaurant column to LA data

la_merged = pd.merge(la_data,la_italian, on='Neighborhood',how='left')


### Clustering the neighborhoods

In [241]:
#Standardized the variables

la_cluster = la_merged[['Population','Median Income','Italian Restaurant']]
la_cluster['Population'] = la_cluster['Population'].astype(str).astype(int)

la_cluster = StandardScaler().fit_transform(la_cluster)


//anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


In [243]:
#cluster the neighborhoods using KMeans 

from sklearn.cluster import KMeans

kclusters = 5

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(la_cluster)

kmeans.labels_[0:10] 

#assign the cluster labels to the neighborhoods in the dataset

la_merged.insert(0, 'Cluster Labels', kmeans.labels_)


array([3, 3, 3, 1, 3, 2, 1, 1, 2, 3], dtype=int32)

In [240]:
la_merged

Cluster Labels                                       Neighborhood    Zip  \
0                3                      90057 - Los Angeles(Westlake)  90057   
1                3  90020 - Los Angeles(Hancock Park, Western Wilt...  90020   
2                3  90005 - Los Angeles(Hancock Park, Koreatown, W...  90005   
3                1  90006 - Los Angeles(Byzantine-Latino Quarter, ...  90006   
4                3                90029 - Los Angeles(East Hollywood)  90029   
5                2  90017 - Los Angeles(Downtown Bunker Hill, City...  90017   
6                1         90011 - Los Angeles(Southeast Los Angeles)  90011   
7                1  90004 - Los Angeles(Hancock Park, Rampart Vill...  90004   
8                2                     90038 - Los Angeles(Hollywood)  90038   
9                3                     90028 - Los Angeles(Hollywood)  90028   
10               3             90037 - Los Angeles(South Los Angeles)  90037   
11               1                         90034 - Los Angeles(Palms)  90034   
12               1        90026 - Los Angeles(Echo Park, Silver Lake)  90026   
13               1  90019 - Los Angeles(Arlington Heights, Country...  90019   
14               1      90044 - Athens,Los Angeles(South Los Angeles)  90044   
15               3  90003 - Los Angeles(South Los Angeles, Southea...  90003   
16               3  90007 - Los Angeles(Southeast Los Angeles, Uni...  90007   
17               3  90018 - Los Angeles(Jefferson Park, Leimert Park)  90018   
18               3                 90033 - Los Angeles(Boyle Heights)  90033   
19               3    90063 - City Terrace,Los Angeles(Boyle Heights)  90063   
20               3  90001 - Los Angeles(South Los Angeles), Floren...  90001   
21               1                           90022 - East Los Angeles  90022   
22               3             90062 - Los Angeles(South Los Angeles)  90062   
23               3    90025 - Los Angeles(Sawtelle, West Los Angeles)  90025   
24               3  90002 - Los Angeles(Southeast Los Angeles, Watts)  90002   
25               1                 90042 - Los Angeles(Highland Park)  90042   
26               3                  90035 - Los Angeles(West Fairfax)  90035   
27               2  90036 - Los Angeles(Fairfax, Melrose, Miracle ...  90036   
28               3                      90024 - Los Angeles(Westwood)  90024   
29               3                    90016 - Los Angeles(West Adams)  90016   
30               0  90013 - Los Angeles(Downtown Central, Downtown...  90013   
31               2  90014 - Los Angeles(Downtown Historic Core, Ar...  90014   
32               3  90059 - Los Angeles(Southeast Los Angeles, Wat...  90059   
33               2  90048 - Los Angeles(Mid-City West), West Holly...  90048   
34               1         90066 - Culver City,Los Angeles(Mar Vista)  90066   
35               3  90043 - Los Angeles(Hyde Park, View Park, Wind...  90043   
36               3             90047 - Los Angeles(South Los Angeles)  90047   
37               3     90032 - Los Angeles(El Sereno, Monterey Hills)  90032   
38               3  90031 - Los Angeles(Lincoln Heights, Montecito...  90031   
39               3  90065 - Los Angeles(Cypress Park, Glassell Par...  90065   
40               3             90061 - Los Angeles(South Los Angeles)  90061   
41               0  90015 - Los Angeles(Dowtown Fashion District, ...  90015   
42               3     90046 - Los Angeles(Hollywood), West Hollywood  90046   
43               3                  90023 - Commerce,East Los Angeles  90023   
44               3  90012 - Los Angeles(Downtown Civic Center, Chi...  90012   
45               3  90008 - Los Angeles(Baldwin Hills, Crenshaw, L...  90008   
46               2                  90067 - Los Angeles(Century City)  90067   
47               2                    90041 - Los Angeles(Eagle Rock)  90041   
48               2  90039 - Los Angeles(Atwater Village, Elysian V...  90039   
49  

In [210]:
#The summary of the clusters

la_sum = la_merged[['Cluster Labels','Population','Median Income','Italian Restaurant']]

la_sum_grouped = la_sum.groupby('Cluster Labels').mean().reset_index()
la_sum_grouped = la_sum_grouped[['Cluster Labels','Population','Median Income','Italian Restaurant']]
la_sum_grouped['Desirability'] = ['Low','Mid-Low','Mid','Mid-High','High']
la_sum_grouped['Number of Neighborhood'] = la_sum.groupby('Cluster Labels').size()
la_sum_grouped

Cluster Labels    Population  Median Income  Italian Restaurant  \
0               0   7204.833333   29296.000000            0.666667   
1               1  70711.700000   48913.900000            0.700000   
2               2  21728.444444   70406.111111            3.000000   
3               3  43232.166667   43308.900000            0.300000   
4               4  22624.200000  112789.600000            0.000000   

  Desirability  Number of Neighborhood  
0          Low                       6  
1      Mid-Low                      10  
2          Mid                       9  
3     Mid-High                      30  
4         High                       5

The cluster with the highest desirability (i.e. the cluster with a high median income and population size but a low number of median income) is the cluster number 4. The table shows that this cluster consists of 5 neighborhoods. The average number of population is 22,624, which is still much lower than that of clusters of 1 and 3. However, the median income of this cluster stands out at $112,789, which is significantly higher than the rest of the clusters. In addition, there is no Italian restaurant around the neighborhoods that form this cluster. The combination of these characteristics leads to the high desirability of this cluster of neighborhood to the company.


In [238]:
#Create a LA map with the clusters

import numpy as np


map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(la_merged['Latitude'], la_merged['Longitude'], la_merged['Neighborhood'], la_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=int(cluster)*3, #more desirable neighborhoods are represented by larger circles
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    

       
map_clusters

In [226]:
#Cluster 4 summary

cluster_4 = la_merged['Cluster Labels'] == 4
la_merged[cluster_4]

Cluster Labels                                       Neighborhood    Zip  \
51               4                             90056 - Ladera Heights  90056   
53               4  90045 - Los Angeles(Los Angeles International ...  90045   
54               4                     90068 - Los Angeles(Hollywood)  90068   
55               4    90049 - Los Angeles(Bel Air Estates, Brentwood)  90049   
58               4  90077 - Los Angeles(Bel Air Estates, Beverly G...  90077   

                                                 Name   Latitude   Longitude  \
51                                     Ladera Heights  33.987945 -118.370442   
53  Los Angeles(Los Angeles International Airport,...  33.954017 -118.402447   
54                             Los Angeles(Hollywood)  34.137411 -118.328915   
55            Los Angeles(Bel Air Estates, Brentwood)  34.091829 -118.491244   
58         Los Angeles(Bel Air Estates, Beverly Glen)  34.102084 -118.451629   

    Population  Median Income  Italian Restaurant  
51        8108          84438                   0  
53       39315          90399                   0  
54       21713          82718                   0  
55       33520         121671                   0  
58       10465         184722                   0

The table above shows the neighborhoods in cluster 4, the most desirable cluster. There is no Italian restaurant within a radius of 1 kilometer from all the 5 neighborhoods in the cluster. The population sizes of these neighborhood range from 8,108 people in Ladera Heights to 39,315 people in the LA International Airport area. Lastly, the median income also varies, from $84,438 for Ladera Heights to 184,722 for Bel Air Estates - Beverly Glen. Overall, all the neighborhoods in this cluster are characterized by the non-existence of nearby Italian restaurants, a modest population size and a high median income.


In [230]:
#Cluster 0 summary

cluster_0 = la_merged['Cluster Labels'] == 0
la_merged[cluster_0]

Cluster Labels                                       Neighborhood    Zip  \
30               0  90013 - Los Angeles(Downtown Central, Downtown...  90013   
41               0  90015 - Los Angeles(Dowtown Fashion District, ...  90015   
52               0  90010 - Los Angeles(Hancock Park, Wilshire Cen...  90010   
56               0                          90040 - Commerce, City of  90040   
57               0  90021 - Los Angeles(Downtown Fashion District,...  90021   
59               0  90058 - Los Angeles(Southeast Los Angeles), Ve...  90058   

                                                 Name   Latitude   Longitude  \
30  Los Angeles(Downtown Central, Downtown Fashion...  34.044639 -118.240413   
41  Los Angeles(Dowtown Fashion District, South Pa...  34.039224 -118.266293   
52  Los Angeles(Hancock Park, Wilshire Center, Win...  34.062125 -118.315709   
56                                  Commerce, City of  33.994524 -118.149953   
57  Los Angeles(Downtown Fashion District, Downtow...  34.029043 -118.239504   
59         Los Angeles(Southeast Los Angeles), Vernon  34.001617 -118.222274   

    Population  Median Income  Italian Restaurant  
30        9727          22808                   1  
41       15134          32979                   1  
52        1943          47115                   1  
56        9798          43585                   0  
57        3003          12864                   1  
59        3624          16425                   0

The table above shows the neighborhood in cluster 0, the least desirable cluster. In 4 out of 6 neighborhoods in the cluster, there is already 1 Italian restaurant with a radius of 1 kilometer from the center. The population size of these neighborhood is also relatively small, with the largest neighborhood having only 15,134 people in it. The median incomes are also significantly smaller compared to those in cluster 4. The highest median income in the cluster is $47,115, which is only slightly higher than half of the lowest median income in cluster 4.


### Conclusion

Based on the results presented above, it is clear that the company should open their restaurants in neighborhoods that make up cluster 4. The reason being these neighborhoods fulfill the criteria set by the company: high median income, decent size of population and less or no competitors. These areas are Ladera Heights, the LA Airport Area, Hollywood, Bel Air Estates - Brentwood and Bel Air Estates - Beverly Glen. These areas are located in either the north-west or south-west part of Los Angeles and are known for having affluent residents.
